src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs

```diff
@@ -0,0 +1,941 @@
+            private int _reserved4;
+            public fixed byte ServerInfo[4]; // Has to be non-zero size, so set it to the Z(4) padding
+            //private int _reserved5;
+        }
+
+        // rfc4178
+        private enum NegotiationToken
+        {
+            NegTokenInit = 0,
+            NegTokenResp = 1
+        }
+
+        private enum NegTokenInit
+        {
+            MechTypes = 0,
+            ReqFlags = 1,
+            MechToken = 2,
+            MechListMIC = 3
+        }
+
+        private enum NegTokenResp
+        {
+            NegState = 0,
+            SupportedMech = 1,
+            ResponseToken = 2,
+            MechListMIC = 3
+        }
+
+        private enum NegState
+        {
+            Unknown = -1,           // Internal. Not in RFC.
+            AcceptCompleted = 0,
+            AcceptIncomplete = 1,
+            Reject = 2,
+            RequestMic = 3
+        }
+
+        internal NTAuthentication(bool isServer, string package, NetworkCredential credential, string? spn, ContextFlagsPal requestedContextFlags, ChannelBinding? channelBinding)
+        {
+            if (isServer)
+            {
+                throw new PlatformNotSupportedException(SR.net_nego_server_not_supported);
+            }
+
+            if (package.Equals("NTLM", StringComparison.OrdinalIgnoreCase))
+            {
+                _isSpNego = false;
+            }
+            else if (package.Equals("Negotiate", StringComparison.OrdinalIgnoreCase))
+            {
+                _isSpNego = true;
+            }
+            else
+            {
+                throw new PlatformNotSupportedException(SR.net_securitypackagesupport);
+            }
+
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, $"package={package}, spn={spn}, requestedContextFlags={requestedContextFlags}");
+
+            if (string.IsNullOrWhiteSpace(credential.UserName) || string.IsNullOrWhiteSpace(credential.Password))
+            {
+                // NTLM authentication is not possible with default credentials which are no-op
+                throw new PlatformNotSupportedException(SR.net_ntlm_not_possible_default_cred);
+            }
+
+            // TODO: requestedContextFlags
+            _credential = credential;
+            _spn = spn;
+            _channelBinding = channelBinding;
+        }
+
+        internal void CloseContext()
+        {
+            // Dispose of the state
+            _negotiateMessage = null;
+            _clientSigningKey = null;
+            _serverSigningKey = null;
+            _clientSeal?.Dispose();
+            _serverSeal?.Dispose();
+            _clientSeal = null;
+            _serverSeal = null;
+            _clientSequenceNumber = 0;
+            _serverSequenceNumber = 0;
+            IsCompleted = true;
+        }
+
+        internal unsafe string? GetOutgoingBlob(string? incomingBlob)
+        {
+            if (IsCompleted)
+            {
+                return null;
```

Comment:
Im wondering if we should add Debug.Assert here. We should not be here if the authentication is completed, right?

Reply:
Yeah, we should. Good point.